ibm-openbmc/bmcweb
# 1110: Implement Fabric PortCollection and Port schemas

1. Implement Fabric PortCollection and Port schemas

This commit implements FabricAdapter PortCollection and Port schemas. PortCollection is a list of associated Ports for a given FabricAdapter.

The association is defined via
- https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/62881.

The backend port examples are also committed via
- https://gerrit.openbmc.org/c/openbmc/openpower-vpd-parser/+/66540
- https://gerrit.openbmc.org/c/openbmc/openpower-vpd-parser/+/70888
- https://gerrit.openbmc.org/c/openbmc/openbmc/+/66541

The current submission only implements the basic properties of Port (e.g. Id, Name etc) as a foundation of the future additional properties.

Tested:
- Redfish Validator passes
- perform GET methods like these:

```
curl -k -X GET https://${bmc}/redfish/v1/Systems/system/FabricAdapters/disk_backplane0
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0",
  "@odata.type": "#FabricAdapter.v1_4_0.FabricAdapter",
  "Id": "disk_backplane0",
  ...
  "Ports": {
    "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports"
  },
  ...
}
```

```
curl -k -X GET https://${bmc}/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports",
  "@odata.type": "#PortCollection.PortCollection",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector5"
    }
  ],
  "Members@odata.count": 2,
  "Name": "Port Collection"
}
```

```
curl -k -X GET https://${bmc}:18080/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4",
  "@odata.type": "#Port.v1_7_0.Port",
  "Id": "dp0_connector4",
  "Name": "dp0_connector4"
}%
```

Also try the invalid port like

```
curl -k -X GET https://${bmc}:18080/redfish/v1/Systems/system/FabricAdapters/io_module1/Ports/INVALID
{
  "error": {
    "@Message.ExtendedInfo": [
      {
        "@odata.type": "#Message.v1_1_1.Message",
        "Message": "The requested resource of type Port named 'INVALID' was not found.",
        "MessageArgs": [
          "Port",
          "INVALID"
        ],
        "MessageId": "Base.1.16.0.ResourceNotFound",
        "MessageSeverity": "Critical",
        "Resolution": "Provide a valid resource identifier and resubmit the request."
      }
    ],
    "code": "Base.1.16.0.ResourceNotFound",
    "message": "The requested resource of type Port named 'INVALID' was not found."
  }
}%
```


Change-Id: I8c64c16764e85c0716e264263708b18f897a2c0c


2. Add Location information for Fabric Port

This commit is to add location information to Port of FabricAdapter. If Port LocationCode property is not found, it will not be shown.

Tested:
  - Redfish Validator passed
  - Check Location from GET Port output

```
% curl -k -X GET https://${bmc}:18080/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4",
  "@odata.type": "#Port.v1_7_0.Port",
  "Id": "dp0_connector4",
  "Location": {
    "PartLocation": {
      "ServiceLabel": "U78DA.ND0.WZS003T-P1-T4"
    }
  },
  "Name": "dp0_connector4"
}
```

Change-Id: Ie015b19612c03a9c656ad14a3f607da04cb4f901


3.  Add Port Status information for Fabric Port

This commit is to add status and health information to Fabric Port like other Redfish resources which implement State/Health and map them to Present/Functional respectfully. State / Health of the Port is useful information for inventory on a GUI/Debug/etc

If the `xyz.openbmc_project.Inventory.Item` interface does not exist, the state status property is set to default "Present".

If the `xyz.openbmc_project.State.Decorator.OperationalStatus` interface does not exist, the health status property is set to default "OK".

Tested:
  - Redfish Validator passed
  - Check status from GET Port output

```
% curl -k -X GET https://${bmc}:18080/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/disk_backplane0/Ports/dp0_connector4",
  "@odata.type": "#Port.v1_7_0.Port",
  "Id": "dp0_connector4",
  "Location": {
    "PartLocation": {
      "ServiceLabel": "U78DA.ND0.WZS003T-P1-T4"
    }
  },
  "Name": "dp0_connector4",
  "Status": {
    "Health": "OK",
    "State": "Enabled"
  }
}
```

Change-Id: Ibb625f2ef1378f77c9520426d2687e305b4f8be5
